### ISSUE TITLE
STDEV aggregation with MSSQL

### ISSUE THREAD

**Issue by SteffenMangold** [aspnet/EntityFrameworkCore#11850]
There is no solution for calling STDEV aggregation function from Entity Core.

**Comment by smitpatel** [comment]
There is no LINQ operator for that. But it would be easy to write a user defined DbFunction to call & translate to STDEV on server.

**Comment by bricelam** [comment]
Should we provide one?
``` vb
Dim stdev = Aggregate b In db.Benchmarks Into EF.Functions.Stdev(b.Result)
```

**Comment by SteffenMangold** [comment]
That would be great 🤗

**Comment by bricelam** [comment]
Notes for triage: PostgreSQL has `stddev`; SQLite has none.

**Comment by smitpatel** [comment]
Re-purposing the issue. Currently there is no support to write a DbFunction which translates to aggregate function over server. Same restriction goes for our method call translators. We need to lift this restriction before we can translate aggregate functions on server.

**Comment by SteffenMangold** [comment]
Is it some kind of a big deal? I try to investigate a realistic timeframe for this. Just to know how to go on with our project. Thanks for your interest so long! ,😊

**Comment by bricelam** [comment]
It's not too bad to work around using value types and FromSql:
``` csharp
class ScalarResult<T>
{
    public T Value { get; set; }
}

class MyContext : DbContext
{
    public DbSet<Benchmark> Benchmarks { get; set; }
    public DbQuery<ScalarResult<double>> Doubles { get; set; }
}
```
``` csharp
var db = new MyContext();
var stdev = Enumerable.Single(
    from r in db.Doubles.FromSql("SELECT STDEV(Result) AS Value FROM Benchmarks")
    select r.Value);
```

**Comment by roji** [comment]
I know this issue has been repurposed, but there seems to definitely be value in having a `EF.Functions.Stdev()` which would be translated to SQL by providers which support it (we have at least SQL Server and PostgreSQL). We can open a different issue to track this.

On a related note, how easy is it for providers to add translation for additional aggregate functions? Regular functions are quite simple (via IMethodCallTranslator) but methods such as `Sum` and `Average` seem to receive some pretty special handling in the core. Is it possible (or is it somehow planned) to add support for translating aggregate functions with the same ease as regular ones?

**Comment by pmiddleton** [comment]
@roji 

Aggregates are treated special by both ReLinq and Core so supporting them will require some extra changes in the backend.  

I will take a look at this after we finish with merging in TVFs. 

@divega -  Should we use this item to track or open another?

**Comment by smitpatel** [comment]
Aggregates are considered ResultOperators in ReLinq and for supporting a single aggregate function means, adding corresponding CustomResultOperator. Registering it in query parser and then translating it in VisitResultOperator method. It is neither easy nor scalable for multiple operators.

It is true that adding function for `Stdev` in EF has value but due to lack of good underlying support, I re-purposed this issue to add back-end support first. Once that is done, I will open another issue to track specific work for Stdev. (or it could be added as proof of concept that pipeline works). My main motivation for re-using same issue was, avoid having multiple tracking issues which talks about adding more functions which would have been essentially blocked on this one.

**Comment by pmiddleton** [comment]
@smitpatel - I was thinking that it might be possible to add a single ResultOperator that can deal with all custom aggregates.  That way we could deal with custom clr aggregate functions in Sql Server also.

If that isn't going to work then just supporting the big hitters should still be possible.  

@roji - does postgres support STDEVP, VAR, VARP?  Those are the ones from Sql Server which are still missing.

**Comment by SteffenMangold** [comment]
Can someone tell me what the process is after "up-for-grabs"? Only to know about what time horizon we are talking. :)

**Comment by divega** [comment]
@SteffenMangold This is the backlog. It means it is definitively not going into 2.1. It may go into the next release if someone (e.g. @pmiddleton has done a lot of great contributions around function support in EF Core) picks it up. Otherwise we may choose to do it anyway in the next release, but given other important features in the backlog or already assigned to the 2.2 milestone, I think the chances are small. 

In the meanwhile my recommendation would be to write the SQL in FromSql(). With the improvements included in 2.1, you should be able to declare a [query type](https://docs.microsoft.com/en-us/ef/core/modeling/query-types) in your model that contains the result of the aggregate and that you can use for context.Query<Result>().FromSql("...").

**Comment by roji** [comment]
@pmiddleton just to be sure, this would include provider-facing infrastructure (similar to today's existing method and member translators), and not just user-facing DbFunction support, right?

**Comment by pmiddleton** [comment]
@roji 

It's all up in the air right now, as I have not started, so we can shoot for whatever we want.  

Are you thinking of an interface for providers to preregister known aggregates so users don't have to.  Similar to how base functions are added in Relational via IMethodProvider?

**Comment by roji** [comment]
That's right. For stuff like `STDDEV` and others, which exist on some providers but not on others, we shouldn't require to use DbFunction but rather have the provider translate something like `EF.Functions.Stddev()`, if at all possible... Ideally something like the existing `SqlServerCompositeMethodCallTranslator`.

**Comment by SteffenMangold** [comment]
Did somebody already picked that task?

**Comment by pmiddleton** [comment]
The query pipeline is being rewritten in 3.0 so this is currently blocked by that work.

**Comment by SteffenMangold** [comment]
Is 3.0 the next major release or is there some other 2.x coming before? Just want to estimate if 1 or 5 years. ^^

**Comment by ajcvickers** [comment]
@SteffenMangold 3.0 is the next major release.

**Comment by smitpatel** [comment]
See #20052

**Comment by roji** [comment]
Just some extra details on aggregate functions...

Some aggregate functions accept some additional non-set arguments. For example, PG `string_agg` concatenates a set of strings, and also accepts a delimiter (`SELECT string_agg(name, ',') FROM users`). Pretty trivial, just mentioning it.

More interesting: there are some cases of functions accepting *multiple* sets; a good example is `jsonb_object_agg`, which reduces two arbitrary sets of values into a single json document:

```sql
postgres=# select jsonb_object_agg('key_' || x::text, x) from
generate_series(1,4) as x;
                 jsonb_object_agg
--------------------------------------------------
 {"key_1": 1, "key_2": 2, "key_3": 3, "key_4": 4}
(1 row)
```
